src/p11_key.c

```diff
@@ -310,6 +308,94 @@ int pkcs11_generate_key(PKCS11_SLOT_private *slot, int algorithm, unsigned int b
 	return 0;
 }
 
+int pkcs11_ec_keygen(PKCS11_SLOT_private *slot, const char *curve,
+		const char *label, unsigned char *id, size_t id_len)
+{
+	PKCS11_CTX_private *ctx = slot->ctx;
+	CK_SESSION_HANDLE session;
+	PKCS11_TEMPLATE pubtmpl = {0}, privtmpl = {0};
+	CK_MECHANISM mechanism = {
+		CKM_EC_KEY_PAIR_GEN, NULL_PTR, 0
+	};
+
+	CK_OBJECT_HANDLE pub_key_obj, priv_key_obj;
+	int rv;
+
+	unsigned char *ecdsa_params = NULL;
+	int ecdsa_params_len = 0;
+	unsigned char *tmp = NULL;
+	ASN1_OBJECT *curve_obj = NULL;
+	int curve_nid = NID_undef;
+
+	if (pkcs11_get_session(slot, 1, &session)) {
+		return -20;
+	}
+
+	curve_nid = EC_curve_nist2nid(curve);
+	if (curve_nid == NID_undef)
+		curve_nid = OBJ_sn2nid(curve);
+	if (curve_nid == NID_undef)
+		curve_nid = OBJ_ln2nid(curve);
+	if (curve_nid == NID_undef)
+		return -21;
+
+	curve_obj = OBJ_nid2obj(curve_nid);
+	if (!curve_obj)
+		return -22;
+	ecdsa_params_len = i2d_ASN1_OBJECT(curve_obj, NULL);
+	ecdsa_params = (unsigned char *)OPENSSL_malloc(ecdsa_params_len);
+	if (!ecdsa_params)
+		return -23;
+	tmp = ecdsa_params;
```

Comment:
Some leftover. Done

Reply:
This is the code I took from the other unmerged EC keygen PR. I will take a deeper look in openssl functions to see what is this magic.

Reply:
So, in line 350, what actually happens is memory space pointed by ecdsa_params gets filled in some way but this also changes ecdsa_params pointer (as a side effect?) which we for some reason don't want and that's why we use a tmp pointer. I don't really understand this cryptic code in openssl:
```
  void ASN1_put_object(unsigned char **pp, int constructed, int length, int tag,
                       int xclass)
  {
      unsigned char *p = *pp;
      int i, ttag;
  
      i = (constructed) ? V_ASN1_CONSTRUCTED : 0;
      i |= (xclass & V_ASN1_PRIVATE);
      if (tag < 31)
          *(p++) = i | (tag & V_ASN1_PRIMITIVE_TAG);
      else {
          *(p++) = i | V_ASN1_PRIMITIVE_TAG;
          for (i = 0, ttag = tag; ttag > 0; i++)
              ttag >>= 7;
          ttag = i;
          while (i-- > 0) {
              p[i] = tag & 0x7f;
              if (i != (ttag - 1))
                  p[i] |= 0x80;
              tag >>= 7;
          }
          p += ttag;                                                         
      }
      if (constructed == 2)
          *(p++) = 0x80;
      else
          asn1_put_length(&p, length);
      *pp = p;
  }
```
and I think spending time to understand what this is doing is not worth it given our current situation. Especially since this seems to work as evidenced by the test, and this behavior only depends on `curve_obj` which is quite straightforward. So I don't anticipate things going wrong wrong here.